src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb

```diff
@@ -0,0 +1,987 @@
+
+                Do
+                    nonTupleTypeChain.Add(namedTypeSymbol)
+                    namedTypeSymbol = CType(namedTypeSymbol.TypeArgumentsNoUseSiteDiagnostics(TupleTypeSymbol.RestPosition - 1), NamedTypeSymbol)
+                Loop While namedTypeSymbol.Arity = TupleTypeSymbol.RestPosition
+
+                If Not namedTypeSymbol.IsTupleType Then
+                    nonTupleTypeChain.Add(namedTypeSymbol)
+                End If
+
+                Debug.Assert(nonTupleTypeChain.Count > 1)
+                tupleCompatibleType = nonTupleTypeChain.Pop()
+
+                Dim typeArgumentsBuilder As ArrayBuilder(Of TypeWithModifiers) = ArrayBuilder(Of TypeWithModifiers).GetInstance(TupleTypeSymbol.RestPosition)
+                Do
+                    Dim extensionTuple As TupleTypeSymbol = TupleTypeSymbol.Create(CType(Nothing, Location), tupleCompatibleType, Nothing, Nothing)
+                    tupleCompatibleType = nonTupleTypeChain.Pop()
+                    tupleCompatibleType = TupleTypeSymbol.ReplaceRestExtensionType(tupleCompatibleType, typeArgumentsBuilder, extensionTuple)
+                Loop While nonTupleTypeChain.Count <> 0
+
+                typeArgumentsBuilder.Free()
+                nonTupleTypeChain.Free()
+            End If
+            Return tupleCompatibleType
+        End Function
+
+        Private Shared Function ReplaceRestExtensionType(tupleCompatibleType As NamedTypeSymbol, typeArgumentsBuilder As ArrayBuilder(Of TypeWithModifiers), extensionTuple As TupleTypeSymbol) As NamedTypeSymbol
+            Dim modifiers As ImmutableArray(Of ImmutableArray(Of CustomModifier)) = Nothing
+            Dim hasTypeArgumentsCustomModifiers As Boolean = tupleCompatibleType.HasTypeArgumentsCustomModifiers
+
+            If hasTypeArgumentsCustomModifiers Then
+                modifiers = tupleCompatibleType.TypeArgumentsCustomModifiers
+            End If
+
+            Dim typeArgumentsNoUseSiteDiagnostics As ImmutableArray(Of TypeSymbol) = tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics
+            typeArgumentsBuilder.Clear()
+
+            For i As Integer = 0 To TupleTypeSymbol.RestPosition - 1 - 1
+                If Not modifiers.IsDefault AndAlso Not modifiers(i).IsDefaultOrEmpty Then
+                    typeArgumentsBuilder.Add(New TypeWithModifiers(typeArgumentsNoUseSiteDiagnostics(i), modifiers(i)))
+                Else
+                    typeArgumentsBuilder.Add(New TypeWithModifiers(typeArgumentsNoUseSiteDiagnostics(i)))
+                End If
+            Next
+
+            If Not modifiers.IsDefault AndAlso Not modifiers(TupleTypeSymbol.RestPosition - 1).IsDefaultOrEmpty Then
+                typeArgumentsBuilder.Add(New TypeWithModifiers(extensionTuple, modifiers(TupleTypeSymbol.RestPosition - 1)))
+            Else
+                typeArgumentsBuilder.Add(New TypeWithModifiers(extensionTuple))
+            End If
```

Comment:
Consider adding a `GetModifiers(modifiers, index)` method that returns `modifiers(i)` or `Nothing` so each of these `Add` calls can be collapsed into a single call: `typeArgumentsBuilder.Add(New TypeWithModifiers(typeArgumentsNoUseSiteDiagnostics(i), GetModifiers(modifiers, i))`.